src/libstore/build/derivation-goal.cc

```diff
@@ -382,7 +382,7 @@ Goal::Co DerivationGoal::gaveUpOnSubstitution()
 
     std::map<ref<const SingleDerivedPath>, GoalPtr, value_comparison> inputGoals;
 
-    if (useDerivation) {
+    if (!drv->inputDrvs.map.empty()) {
```

Comment:
Technically we don't even need the condition, as the loop below will be run zero times regardless.